### PR TITLE
[ML] Instrument the regression code for memory usage and training run time

### PR DESCRIPTION
**Pull request by tveasey**
This instruments the regression code to extract peak memory usage and training run time. This enabled me to test memory usage estimation better and I've fixed a bug where we weren't properly accounting for the maximum tree and forest size.

### CODE REVIEW

File: include/maths/CBoostedTreeImpl.h [hunk 1]
```diff
@@ -402,17 +412,29 @@ class MATHS_EXPORT CBoostedTreeImpl final {
...
+                                               std::size_t numberCols,
+                                               double featureBagFraction,
+                                               std::size_t numberSplitsPerFeature) {
             std::size_t featureBagSize{
-                static_cast<std::size_t>(ceil(featureBagFraction) * (numberCols - 1)) *
+                static_cast<std::size_t>(std::ceil(
+                    featureBagFraction * static_cast<double>(numberCols - 1))) *
                 sizeof(std::size_t)};
-            std::size_t rowMaskSize{numberRows / 32};
+            std::size_t rowMaskSize{numberRows / 256};
```
Comment (dimitris-athanasiou) [inline include/maths/CBoostedTreeImpl.h]: Could we document this magic number?

File: lib/api/CDataFrameBoostedTreeRunner.cc [hunk 2]
```diff
@@ -145,10 +156,19 @@ void CDataFrameBoostedTreeRunner::runImpl(const TStrVec& featureNames,
         return;
     }
 
+    core::CProgramCounters::counter(counter_t::E_DFTPMEstimatedPeakMemoryUsage) =
+        this->estimateMemoryUsage(frame.numberRows(),
+                                  frame.numberRows() / this->numberPartitions(),
+                                  frame.numberColumns());
+
+    core::CStopWatch watch{true};
```
Comment (dimitris-athanasiou) [inline lib/api/CDataFrameBoostedTreeRunner.cc]: Should we not also add a watch tracking outlier detection duration?

Comment (tveasey) [reply lib/api/CDataFrameBoostedTreeRunner.cc]: This is done at a lower level for outlier detection: see for example [this](https://github.com/elastic/ml-cpp/blob/master/lib/maths/COutliers.cc#L895) line.

File: include/maths/CBoostedTreeImpl.h (hunk 1 above)
Comment (tveasey) [reply include/maths/CBoostedTreeImpl.h]: See [68158b7](https://github.com/elastic/ml-cpp/pull/607/commits/68158b7f586e39cdd6e86b205ea9efab69b02d7e).

**Comment by tveasey** [comment]
@dimitris-athanasiou I addressed your two comments. Could you take another look?

**Pull request merged**